include/mixer.h

```diff
@@ -0,0 +1,461 @@
+ * decide to ignore the "wpos" argument and simply continue decoding the audio
+ * from the place were it last stopped. If "seeking" is true, instead, it means
+ * that there was a jump in position that should be taken into account.
+ *
+ * Normally, a seeking is required in the following situations:
+ *
+ *   * At start (#mixer_ch_play): the first read issued by the mixer will
+ *     be at position 0 with seeking == true (unless the waveform was
+ *     already partially cached in the sample buffer's channel).
+ *   * At user request (#mixer_ch_set_pos).
+ *   * At loop: if a loop is specified in the waveform (loop_len != 0), the
+ *     mixer will seek when requires to execute the loop.
+ *
+ * Notice that producing more samples than requested in "wlen" might break
+ * the 8-byte buffer alignment guarantee that #samplebuffer_append tries to
+ * provide. For instance, if the read function is called requesting 24 samples,
+ * but it produces 25 samples instead, the alignment in the buffer will be lost,
+ * and next call to samplebuffer_append will return an unaligned pointer.
+ *
+ * "ctx" is an opaque pointer that is provided as context to the function, and
+ * is specified in the waveform.
+ */
+typedef void (*WaveformRead)(void *ctx, samplebuffer_t *sbuf, int wpos, int wlen, bool seeking);
+
+/**
+ * @brief A waveform that can be played back through the mixer.
+ * 
+ * waveform_t represents a waveform that can be played back by the mixer.
+ * A waveform_t does not hold the actual samples because most real-world use
+ * cases do not keep all samples in memory, but rather load them and/or
+ * decompress them in real-time while the playback is happening.
+ * So waveform_t instead should be thought of as the generator of a
+ * waveform.
+ *
+ * To create a waveform, use one of waveform implementations such as wav64. 
+ * Waveform implementations are in charge of generating the samples by actually
+ * implementing an audio format like VADPCM or MPEG-2.
+ *
+ * Waveforms can produce samples as 8-bit or 16-bit. Samples must always be
+ * signed. Stereo waveforms are currently not supported. If an audio format
+ * encodes two channels for stereo playback, the player should expose two
+ * different waveforms that must be played in two different mixer channels.
+ */
+typedef struct waveform_s {
+    // Name of the waveform (for debugging purposes)
+    const char *name;
+
+    // Width of a sample of this waveform, in bits. Supported values are 8 or 16.
+    // Notice that samples must always be signed.
+    uint8_t bits;
+
+    // Number of interleaved audio channels in this waveforms. Currently, this
+    // must always be 1.
+    uint8_t channels;
+
+    // Desired playback frequency (in samples per second, aka Hz).
+    float frequency;
+
+    // Length of the waveform, in number of samples. If the length is not
+    // known, this value can be initialized to a very large number (eg: 2**31-1)
```

Comment:
As discussed on discord, 2^31-1 does not work, currently the max len appears to be 0x40000 - 1.

Reply:
I've pushed a change that allows for longer waveforms to be played back, switching to 64-bit positions in C code while keeping RSP 32-bit (RSP just processes a short window at a time anyway).

Reply:
Thanks! Appears to work.
I changed my code to WAVEFORM_UNKNOWN_LEN with loop_len = 0 and had it running for the past 10 mins or so with no issues in the music stream.